cylc/flow/parsec/fileparse.py

```diff
@@ -239,27 +268,63 @@ def read_and_proc(fpath, template_vars=None, viewcfg=None, asedit=False):
         flines = inline(
             flines, fdir, fpath, False, viewcfg=viewcfg, for_edit=asedit)
 
+    template_vars['CYLC_VERSION'] = __version__
+
     # process with EmPy
     if do_empy:
+        if (
+            extra_vars['templating detected'] == 'empy:suite.rc' and
+            not re.match(r'^#![Ee]m[Pp]y\s*', flines[0])
+        ):
+            if not re.match(r'^#!', flines[0]):
+                flines.insert(0, '#!empy')
+            else:
+                raise FileParseError(
+                    "Your file has shebang line incompatible with empy"
+                    " variables."
+                )
         if flines and re.match(r'^#![Ee]m[Pp]y\s*', flines[0]):
             LOG.debug('Processing with EmPy')
+            if extra_vars['templating detected'] == 'empy:suite.rc':
+                for key, value in template_vars.items():
+                    extra_vars[key] = value
+                template_vars = extra_vars
             try:
                 from cylc.flow.parsec.empysupport import empyprocess
             except (ImportError, ModuleNotFoundError):
                 raise ParsecError('EmPy Python package must be installed '
                                   'to process file: ' + fpath)
-            flines = empyprocess(flines, fdir, template_vars)
+            flines = empyprocess(
+                flines, fdir, template_vars
+            )
 
     # process with Jinja2
     if do_jinja2:
+        if (
+            extra_vars['templating detected'] == 'jinja2:suite.rc' and
+            not re.match(r'^#![jJ]inja2\s*', flines[0])
+        ):
+            if not re.match(r'^#!', flines[0]):
+                flines.insert(0, '#!jinja2')
+            else:
+                raise FileParseError(
+                    "Your file has shebang line incompatible with jinja2"
+                    " variables."
+                )
         if flines and re.match(r'^#![jJ]inja2\s*', flines[0]):
             LOG.debug('Processing with Jinja2')
+            if extra_vars['templating detected'] == 'jinja2:suite.rc':
+                for key, value in template_vars.items():
+                    extra_vars['template variables'][key] = value
```

Comment:
This brings about another problem: `--set` variables are always strings, while rose vars in here can be of any type allowed by `literal_eval`, so a blanket override like this might change the expected type.

Reply:
Yes and it's exactly the same problem again, we have but two options:

1. Allow the override, the user got exactly what they asked for.
2. Fail, reject the override, it might not be safe.

Reply:
> Yes and it's exactly the same problem again, we have but two options:
> 
> 1. Allow the override, the user got exactly what they asked for.
> 2. Fail, reject the override, it might not be safe.
> 
> Unless there are any better ideas I'm inclined to go with (1) for this and all other previously outlined cases, albeit with a warning.

3. Change the semantics of cylc `--set` vars to go via `literal_eval` to match the semantics of rose vars.

This would introduce incompatibility from the current behaviour, but would be better in the long run anyway IMO. Porting from `--set foo=bar` to `--set foo="bar"` is trivial, similarly for the `--set-file` option and there's no fixed-name equivalent of `rose-suite.conf` to worry about.

Reply:
Erm, that's orthogonal, you still have to pick either (1) or (2), changes to the way Cylc interprets CLI inputs is a different kettle of fish.

Reply:
Discussed at the monthly meeting, considering other CLI breakages we are ok with this, if it is a problem then we could potentially use a separate CLI flag for typed as opposed to string options.

I've raised a PR here: https://github.com/cylc/cylc-flow/pull/3974

Note that it upgrades template var set files in the same way which we may or may not want to go ahead with depending on usage.

Reply:
Suggest going with (1) and just raising a warning something along the lines of:

```python
if key in extra_vars['template variables']:
    LOG.warning(f'overriding {key}: {extra_vars["template variables'']} -> {value}')
```